### ISSUE TITLE
ms_deform_attn_forward_cuda" not implemented for 'BFloat16

### ISSUE THREAD

**Issue by ChinChyi** [IDEA-Research/Grounded-SAM-2#38]
Hello!

This is the problem when I use `grounded_sam2_local_demo.py` for image inference

**Comment by rentainhe** [comment]
Hi @ChinChyi 

This is caused by `Deformable Attention` operator, which did not support `BFloat16` inference, we will fix this bug later

**Comment by rentainhe** [comment]
Hi @ChinChyi 

Have you change any code in your local env, we have fix this bug in our original implementation here:

https://github.com/IDEA-Research/Grounded-SAM-2/blob/6e0ddadf7cf8f33610ba7f6fdce30673526afa0c/grounded_sam2_local_demo.py#L53

By removing

```python
# FIXME: figure how does this influence the G-DINO model
torch.autocast(device_type="cuda", dtype=torch.bfloat16).__enter__()

if torch.cuda.get_device_properties(0).major >= 8:
    # turn on tfloat32 for Ampere GPUs (https://pytorch.org/docs/stable/notes/cuda.html#tensorfloat-32-tf32-on-ampere-devices)
    torch.backends.cuda.matmul.allow_tf32 = True
    torch.backends.cudnn.allow_tf32 = True
```

After we get grounding results

**Comment by Khlann** [comment]
So, what is the solution?I have also encounter this problem.Thank you very much!

**Comment by Khlann** [comment]
This error happened when I called` self.sam2_predictor.predict`twice

**Comment by rentainhe** [comment]
> This error happened when I called` self.sam2_predictor.predict`twice

Would you like to share your code with us which may be more convenient for us to debug this issue.

**Comment by ChinChyi** [comment]
> Hi @ChinChyi
> 
> Have you changed any code in your local env, we have fix this bug in our original implementation here:
> 
> https://github.com/IDEA-Research/Grounded-SAM-2/blob/6e0ddadf7cf8f33610ba7f6fdce30673526afa0c/grounded_sam2_local_demo.py#L53
> 
> By removing
> 
> ```python
> # FIXME: figure how does this influence the G-DINO model
> torch.autocast(device_type="cuda", dtype=torch.bfloat16).__enter__()
> 
> if torch.cuda.get_device_properties(0).major >= 8:
>     # turn on tfloat32 for Ampere GPUs (https://pytorch.org/docs/stable/notes/cuda.html#tensorfloat-32-tf32-on-ampere-devices)
>     torch.backends.cuda.matmul.allow_tf32 = True
>     torch.backends.cudnn.allow_tf32 = True
> ```
> 
> After running grounding dino

Thanks

**Comment by MagdalenaKotynia** [comment]
@ChinChyi changing `bloat16` to `float16`  (`torch.autocast(device_type="cuda", dtype=torch.bfloat16).__enter__()` to `torch.autocast(device_type="cuda", dtype=torch.float16).__enter__()` ) helped for me.

**Comment by Xqq2620xx** [comment]
> This error happened when I called` self.sam2_predictor.predict`twice

I meet the same bug. Because `torch.autocast(device_type="cuda", dtype=torch.bfloat16).__enter__() ` inside the loop but did not properly exit` (__exit__)`, causing an autocast state leak.
Then autocast automatically converts torch.Tensor data types, but the ms_deform_attn_forward_cuda operator does not support bfloat16, leading to an error in the second iteration.
Just Use `with torch.autocast(...)` inside the loop to ensure that autocast only applies to a single inference per iteration. It works for me:
`# FIXME: figure how does this influence the G-DINO model
        with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
            if torch.cuda.get_device_properties(0).major >= 8:
                torch.backends.cuda.matmul.allow_tf32 = True
                torch.backends.cudnn.allow_tf32 = True


            masks, scores, logits = sam2_predictor.predict(
                point_coords=None,
                point_labels=None,
                box=input_boxes,
                multimask_output=False,
            )

            """
            Post-process the output of the model to get the masks, scores, and logits for visualization
            """
            ......
`
